frontend/src/components/P4Editor.tsx

```diff
@@ -3,64 +3,200 @@ import Button from "@material-ui/core/Button";
 import P4LanguageServiceEditor from './P4LanguageServiceEditor'
 import * as monaco from 'monaco-editor';
 import APIRequest from '../api/Request'
+import Snackbar from '@material-ui/core/Snackbar';
+import MuiAlert, { AlertProps } from '@material-ui/lab/Alert';
+import { Box, ButtonGroup } from "@material-ui/core";
+import CloudUploadIcon from '@material-ui/icons/CloudUpload';
+import CloudDownloadIcon from '@material-ui/icons/CloudDownload';
+import Dialog from '@material-ui/core/Dialog';
+import DialogActions from '@material-ui/core/DialogActions';
+import DialogContent from '@material-ui/core/DialogContent';
+import DialogContentText from '@material-ui/core/DialogContentText';
+
+type Severity = "error" | "success" | "info" | "warning" | undefined;
+
 interface State {
   code: string
+  editorResult: string
+  editorSeverity: Severity
+  editorNotificationOpen: boolean
+  fileChanged: boolean
+  editorConfirmationDialogOpen: boolean
 }
 
 interface P4EditorProps {
   endpoint: string;
 }
 
+function Alert(props: JSX.IntrinsicAttributes & AlertProps) {
+  return <MuiAlert elevation={6} variant="filled" {...props} />;
+}
+
 export default class P4Editor extends React.Component<P4EditorProps> {
   public state: State;
   private editor!: monaco.editor.IStandaloneCodeEditor;
-  
+
   constructor(props: P4EditorProps) {
     super(props);
     this.state = {
       code: "",
+      editorResult: "",
+      editorSeverity: "error",
+      editorNotificationOpen: false,
+      fileChanged: false,
+      editorConfirmationDialogOpen: false
     };
     this.save = this.save.bind(this);
+    this.load = this.load.bind(this);
   }
 
-  editorDidMount(editor: monaco.editor.IStandaloneCodeEditor){
+  editorDidMount(editor: monaco.editor.IStandaloneCodeEditor) {
     editor.focus();
     this.editor = editor;
-    fetch(APIRequest(`${this.props.endpoint}`, {headers: {'Content-Type': 'application/json', authorization: localStorage.getItem("token") || ""} }))
+    fetch(APIRequest(`${this.props.endpoint}`, { headers: { 'Content-Type': 'application/json', authorization: localStorage.getItem("token") || "" } }))
       .then((response) => response.text())
       .then((data) => {
         this.setState({ code: data });
       })
       .catch((err) => console.error(err));
   };
 
-  onChange(newValue: string){
-    this.setState({code: newValue})
+  onChange(newValue: string) {
+    this.setState({ code: newValue })
+    this.setState({ fileChanged: true })
   };
 
-  save(): void {
-    fetch(APIRequest(this.props.endpoint, {
-      method: "post", 
-    body: this.editor.getModel()?.getValue(),headers: {'Content-Type': 'text/plain', 
-    authorization: localStorage.getItem("token") || ""}  }))
-      .then((response) => response.text())
-      .catch((error) => {
-        console.error(error);
-      });
+  async save(): Promise<void> {
+    this.setState({ fileChanged: false })
```

Comment:
you can use one object for setState instead of splitting it up into 4 functions since this is not a hook

Reply:
Fixed! Seems to look way more ellegant ;)